diff --git a/src/streamsvr.c b/src/streamsvr.c
index 19733b0..228b6c2 100644
--- a/src/streamsvr.c
+++ b/src/streamsvr.c
@@ -490,7 +490,7 @@ static void *strsvrthread(void *arg)
     sol_t sol_nmea={{0}};
     uint32_t tick,tick_nmea;
     uint8_t buff[1024];
-    int i,n,cyc;
+    int i,n,cyc,nsum;
     
     tracet(3,"strsvrthread:\n");
     
@@ -499,10 +499,25 @@ static void *strsvrthread(void *arg)
     
     for (cyc=0;svr->state;cyc++) {
         tick=tickget();
+        nsum=0;
         
+        /* read message from output stream if connected */
+        for (i=1;i<svr->nstr;i++) {
+            while (strstat(svr->stream+i,NULL)>=2 &&
+                  (n=strread(svr->stream+i,buff,sizeof(buff)))>0) {
+                nsum+=n;
+                /* relay back message from output stream to input stream */
+                if (i==svr->relayback) {
+                    strwrite(svr->stream,buff,n);
+                }
+                /* write data to log stream */
+                strwrite(svr->strlog+i,buff,n);
+            }
+        }
+
         /* read data from input stream */
         while ((n=strread(svr->stream,svr->buff,svr->buffsize))>0&&svr->state) {
-            
+            nsum+=n;
             /* write data to output streams */
             for (i=1;i<svr->nstr;i++) {
                 if (svr->conv[i-1]) {
@@ -521,19 +536,6 @@ static void *strsvrthread(void *arg)
             }
             rtklib_unlock(&svr->lock);
         }
-        for (i=1;i<svr->nstr;i++) {
-            
-            /* read message from output stream if connected */
-            while (strstat(svr->stream+i,NULL)>=2 &&
-                  (n=strread(svr->stream+i,buff,sizeof(buff)))>0) {
-                /* relay back message from output stream to input stream */
-                if (i==svr->relayback) {
-                    strwrite(svr->stream,buff,n);
-                }
-                /* write data to log stream */
-                strwrite(svr->strlog+i,buff,n);
-            }
-        }
         /* write periodic command to input stream */
         for (i=0;i<svr->nstr;i++) {
             periodic_cmd(cyc*svr->cycle,svr->cmds_periodic[i],svr->stream+i);
@@ -547,7 +549,8 @@ static void *strsvrthread(void *arg)
             strsendnmea(svr->stream,&sol_nmea);
             tick_nmea=tick;
         }
-        sleepms(svr->cycle-(int)(tickget()-tick));
+        if (!nsum)
+           sleepms(svr->cycle-(int)(tickget()-tick));
     }
     for (i=0;i<svr->nstr;i++) strclose(svr->stream+i);
     for (i=0;i<svr->nstr;i++) strclose(svr->strlog+i);
@@ -761,10 +764,10 @@ extern void strsvrstop(strsvr_t *svr, char **cmds)
 extern void strsvrstat(strsvr_t *svr, int *stat, int *log_stat, int *byte,
                        int *bps, char *msg)
 {
-    char s[MAXSTRMSG]="",*p=msg;
     int i,bps_in;
     
     tracet(4,"strsvrstat:\n");
+    *msg=0;
     
     for (i=0;i<svr->nstr;i++) {
         if (i==0) {
@@ -773,9 +776,9 @@ extern void strsvrstat(strsvr_t *svr, int *stat, int *log_stat, int *byte,
         else {
             strsum(svr->stream+i,NULL,&bps_in,byte+i,bps+i);
         }
-        stat[i]=strstat(svr->stream+i,s);
-        if (*s) p+=sprintf(p,"(%d) %s ",i,s);
-        log_stat[i]=strstat(svr->strlog+i,s);
+
+        stat[i]=strstat(svr->stream+i,*msg?NULL:msg);
+        log_stat[i]=strstat(svr->strlog+i,NULL);
     }
 }
 /* peek input/output stream ----------------------------------------------------
