diff --git a/tools/install.sh b/tools/install.sh
index a3972d5..f0c7454 100755
--- a/tools/install.sh
+++ b/tools/install.sh
@@ -289,7 +289,7 @@ install_rtkbase_bundled() {
     ARCHIVE=$(awk '/^__ARCHIVE__/ {print NR + 1; exit 0; }' "${0}")
     # Check if there is some content after __ARCHIVE__ marker (more than 100 lines)
     [[ $(sed -n '/__ARCHIVE__/,$p' "${0}" | wc -l) -lt 100 ]] && echo "RTKBASE isn't bundled inside install.sh. Please choose another source" && exit 1  
-    sudo -u "${RTKBASE_USER}" tail -n+${ARCHIVE} "${0}" | tar xpJv && \
+    sudo -u "${RTKBASE_USER}" tail -n+${ARCHIVE} "${0}" | sudo -u "${RTKBASE_USER}" tar xpJv && \
     _add_rtkbase_path_to_environment
 }
 
@@ -337,7 +337,7 @@ rtkbase_requirements(){
         cp "${rtkbase_path}/settings.conf.default" "${rtkbase_path}/settings.conf"
       fi
       #Then launch check cpu temp script for OPI zero LTS
-      source "${rtkbase_path}/tools/opizero_temp_offset.sh"
+      #source "${rtkbase_path}/tools/opizero_temp_offset.sh"
       #venv module installation
       sudo -u "${RTKBASE_USER}" "${python_venv}" -m pip install --upgrade pip setuptools wheel  --extra-index-url https://www.piwheels.org/simple
       # install prebuilt wheel for cryptography because it is unavailable on piwheels (2023/01)
