diff --git a/run_cast.sh b/run_cast.sh
index db7a656..ae5a710 100755
--- a/run_cast.sh
+++ b/run_cast.sh
@@ -8,16 +8,26 @@ BASEDIR="$(dirname "$0")"
 source <( grep -v '^#' "${BASEDIR}"/settings.conf | grep '=' ) #import settings
 
 receiver_info="RTKBase ${receiver},${version} ${receiver_firmware}"
-in_serial="serial://${com_port}:${com_port_settings}#${receiver_format}"
-in_tcp="tcpcli://localhost:${tcp_port}#${receiver_format}"
+in_serial="serial://${com_port}:${com_port_settings}#${receiver_format} -s 60000 -d 100 -r 1000"
+in_tcp="tcpcli://localhost:${tcp_port}#${receiver_format} -s 60000 -d 100 -r 1000"
 #in_ext_tcp is mainly for dev purpose to receive a raw stream from another base
 in_ext_tcp="tcpcli://${ext_tcp_source}:${ext_tcp_port}#${receiver_format}"
 
-out_caster_A="ntrips://:${svr_pwd_a}@${svr_addr_a}:${svr_port_a}/${mnt_name_a}#rtcm3 -msg ${rtcm_msg_a} -p ${position}"
+if [[ ${mnt_name_a} == TCP ]] && [[ ${svr_pwd_a} == TCP ]]
+  then
+    out_caster_A="tcpcli://${svr_addr_a}:${svr_port_a}#rtcm3 -msg ${rtcm_msg_a} -p ${position}"
+  else
+    out_caster_A="ntrips://:${svr_pwd_a}@${svr_addr_a}:${svr_port_a}/${mnt_name_a}#rtcm3 -msg ${rtcm_msg_a} -p ${position}"
+fi
 #add receiver options if it exists
 [[ ! -z "${ntrip_a_receiver_options}" ]] && out_caster_A=""${out_caster_A}" -opt "${ntrip_a_receiver_options}""
 
-out_caster_B="ntrips://:${svr_pwd_b}@${svr_addr_b}:${svr_port_b}/${mnt_name_b}#rtcm3 -msg ${rtcm_msg_b} -p ${position}"
+if [[ ${mnt_name_b} == TCP ]] && [[ ${svr_pwd_b} == TCP ]]
+  then
+    out_caster_B="tcpcli://${svr_addr_b}:${svr_port_b}#rtcm3 -msg ${rtcm_msg_b} -p ${position}"
+  else
+    out_caster_B="ntrips://:${svr_pwd_b}@${svr_addr_b}:${svr_port_b}/${mnt_name_b}#rtcm3 -msg ${rtcm_msg_b} -p ${position}"
+fi
 #add receiver options if it exists
 [[ ! -z "${ntrip_b_receiver_options}" ]] && out_caster_B=""${out_caster_B}" -opt "${ntrip_b_receiver_options}""
 
@@ -28,15 +38,15 @@ if [[ ${local_ntripc_user} == '' ]] && [[ ${local_ntripc_pwd} == '' ]]
   else
     local_ntripc_auth='B' #Basic authentification
 fi
-out_local_caster_source_table="${local_ntripc_mnt_name};rtcm3;${local_ntripc_msg};${receiver_frequency_count};GPS+GLO+GAL+BDS+QZS;NONE;NONE;${array_pos[0]};${array_pos[1]};0;0;RTKBase_${receiver},${version};NONE;${local_ntripc_auth};N;;"
+out_local_caster_source_table="${local_ntripc_mnt_name};rtcm3;${local_ntripc_msg};${receiver_frequency_count};GPS+GLO+GAL+BDS+QZS+SBAS+IRN;NONE;NONE;${array_pos[0]};${array_pos[1]};0;0;RTKBase_${receiver},${version};NONE;${local_ntripc_auth};N;;"
 out_local_caster="ntripc://${local_ntripc_user}:${local_ntripc_pwd}@:${local_ntripc_port}/${local_ntripc_mnt_name}:${out_local_caster_source_table}#rtcm3 -msg ${local_ntripc_msg} -p ${position}"
 #add receiver options if it exists
 [[ ! -z "${local_ntripc_receiver_options}" ]] && out_local_caster="${out_local_caster} -opt ${local_ntripc_receiver_options}"
-out_tcp="tcpsvr://:${tcp_port}"
+out_tcp="tcpsvr://localhost:${tcp_port}"
 
 out_file="file://${datadir}/${file_name}.${receiver_format}::T::S=${file_rotate_time} -f ${file_overlap_time}"
 
-out_rtcm_svr="tcpsvr://:${rtcm_svr_port}#rtcm3 -msg ${rtcm_svr_msg} -p ${position}"
+out_rtcm_svr="tcpsvr://:${rtcm_svr_port} -b 1 -msg ${rtcm_svr_msg} -p ${position}"
 #add receiver options if it exists
 [[ ! -z "${rtcm_receiver_options}" ]] && out_rtcm_svr=""${out_rtcm_svr}" -opt "${rtcm_receiver_options}""
 
@@ -56,6 +66,22 @@ mkdir -p ${logdir}
 
   case "$2" in
     out_tcp)
+    #echo receiver="${receiver}" position="${position}"
+    # For Unicore reciever setup base position to receiver (if it's not zero)
+    if [[ "${receiver}" =~ Unicore* ]] || [[ "${receiver}" =~ Bynav ]] || [[ "${receiver}" =~ Septentrio ]]
+    then
+       #echo ${BASEDIR}/UnicoreSetBasePos.sh ${com_port}:${com_port_settings%%:*} "${position}" "${receiver}" "${antenna_info}"
+       ${BASEDIR}/UnicoreSetBasePos.sh ${com_port} ${com_port_settings%%:*} "${position}" "${receiver}" "${antenna_info}"
+       #${BASEDIR}/UnicoreSetBasePos.sh ${com_port} ${com_port_settings%%:*} "${position}" "${receiver}" "${antenna_info}">>${BASEDIR}/debug.log 2>&1
+       exitcode=$?
+       #echo UnicoreSetBasePos exitcode=${exitcode} >>${BASEDIR}/debug.log 2>&1
+       #echo ============================================== >>${BASEDIR}/debug.log 2>&1
+       if [[ ${exitcode} != 0 ]]
+       then
+          echo run_cast exit with exitcode ${exitcode}
+          exit ${exitcode}
+       fi
+    fi
     #echo ${cast} -in ${!1} -out $out_tcp
     # What is this ${!1} ? It's variable indirection
     ${cast} -in ${!1} -out ${out_tcp} -b 1 -t ${level} -fl ${logdir}/str2str_tcp.log &
@@ -63,11 +89,11 @@ mkdir -p ${logdir}
 
   out_caster_A)
     #echo ${cast} -in ${!1} -out $out_caster
-    ${cast} -in ${!1} -out ${out_caster_A} -i "${receiver_info}" -a "${antenna_info}" -t ${level} -fl ${logdir}/str2str_ntrip_A.log &
+    ${cast} -in ${!1} -out ${out_caster_A} -i "${receiver_info}" -a "${antenna_info}" -t ${level} -fl ${logdir}/str2str_ntrip_A.log -fg ${BASEDIR}/ntrip_A.flg &
     ;;
 
   out_caster_B)
-    ${cast} -in ${!1} -out ${out_caster_B} -i "${receiver_info}" -a "${antenna_info}" -t ${level} -fl ${logdir}/str2str_ntrip_B.log &
+    ${cast} -in ${!1} -out ${out_caster_B} -i "${receiver_info}" -a "${antenna_info}" -t ${level} -fl ${logdir}/str2str_ntrip_B.log -fg ${BASEDIR}/ntrip_B.flg &
     ;;
 
   out_local_caster)
