diff --git a/web_app/static/settings.js b/web_app/static/settings.js
index c63e6e3..bb84b7a 100644
--- a/web_app/static/settings.js
+++ b/web_app/static/settings.js
@@ -347,8 +347,19 @@ $(document).ready(function () {
         if (response['result'] === 'success') {
             detectBodyElt.innerHTML = '<b>' + response['gnss_type'] + '</b>' + ' detected on ' + '<b>' + response['port'] + '</b>' + '<br>' + '<br>' + 'Do you want to apply?';
             detectApplyBtnElt.onclick = function (){
+                socket.emit("detection_apply", msg);
                 document.querySelector('#com_port').value = response['port'].replace(/^\/dev\//, '');
                 document.querySelector('#com_port_settings').value = response['port_speed'] + ':8:n:1';
+                document.querySelector('#receiver').value = response['gnss_type'];
+                var recvformat;
+                if (response['gnss_type'].search(/u-blox/) != -1) {
+                   recvformat = "ubx"
+                // } else if (response['gnss_type'].search(/Septentrio/) != -1) {
+                //   recvformat = "sbf"
+                } else {
+                   recvformat = "rtcm3"
+                }
+                document.querySelector('#receiver_format').value = recvformat;
                 // NEW METHOD from https://stackoverflow.com/questions/35154348/trigger-form-submission-with-javascript
                 document.getElementById("main").dispatchEvent(new SubmitEvent('submit', {cancelable: true}));
                 if (response['then_configure']) {
@@ -703,6 +714,15 @@ $(document).ready(function () {
           };
       };
 
+    // ########################### HANDLE TAILSCALE ###########################
+
+    document.getElementById('tailscale-button').onclick = function (){
+        socket.emit("get tailscale href");
+    }
+    socket.on("tailscale_href", function(msg) {
+        window.open(msg, '_blank');
+    })
+
     // ####################### HANDLE REBOOT & SHUTDOWN #######################
 
     $("#reboot-button").on("click", function() {
@@ -735,5 +755,28 @@ $(document).ready(function () {
         socket.emit("shutdown device");
     })
 
+    $("#upgrade-button").on("click", function() {
+        $("#upgradeModal .modal-body").html('Update & upgrade linux. Are you sure?');
+        $("#confirm-upgrade-button").prop("hidden", false);
+        $("#confirm-upgrade-button").prop("disabled", false);
+        $("#confirm-upgrade-button").html('Upgrade');
+        $("#upgrade-cancel-button").prop("hidden", false);
+        $("#upgrade-cancel-button").prop("disabled", false);
+        $("#upgrade-cancel-button").html('Cancel');
+        $("#upgradeModal").modal();
+    })
+    $("#confirm-upgrade-button").on("click", function() {
+        $("#upgradeModal .modal-body").html('<div class="align-items-center">Upgrading now...  <div class="spinner-border ml-auto" role="status" aria-hidden="true"></div></div>');
+        $(this).html('<span class="spinner-border spinner-border-sm"></span> Upgrading...');
+        $(this).prop("disabled", true);
+        $("#upgrade-cancel-button").prop("hidden", true);
+        socket.emit("update and upgrade");
+    })
+    socket.on("upgrade_finished", function(msg) {
+        $("#upgradeModal .modal-body").html(msg);
+        $("#confirm-upgrade-button").prop("hidden", true);
+        $("#upgrade-cancel-button").prop("hidden", false);
+        $("#upgrade-cancel-button").html('Continue');
+    })
     // end of document.ready
 });
