diff --git a/web_app/static/status.js b/web_app/static/status.js
index 6f7cd6e..36cad8b 100644
--- a/web_app/static/status.js
+++ b/web_app/static/status.js
@@ -191,12 +191,28 @@ $(document).ready(function () {
 
         //update map marker position
         // TODO refactoring with the same instructions in graph.js
-        var coordinates = (typeof(msg['pos llh single (deg,m) rover']) == 'undefined') ? '000' : msg['pos llh single (deg,m) rover'].split(',');
+        var llh = msg['pos llh single (deg,m) rover'];
+        if (typeof(llh) != 'undefined') {
+           if (llh != '0.00000000,0.00000000,0.000') {
+              var coordinates = llh.split(',');
+
+              var lat_value = coordinates[0].substring(0, 11) + Array(11 - coordinates[0].substring(0, 11).length + 1).join(" ");
+              var lon_value = coordinates[1].substring(0, 11) + Array(11 - coordinates[1].substring(0, 11).length + 1).join(" ");
+              var height_value = coordinates[2].substring(0, 11) + Array(11 - coordinates[2].substring(0, 11).length + 1 + 2).join(" ");
+
+              var lat_number = Number(lat_value);
+              var lon_number = Number(lon_value);
+              var height_value = Number(height_value);
+
+              if ((lat_number == 0.0) && (lon_number == 0.0) && (height_number == 0.0))
+                 console.log('position is zero');
+              else
+                 locMark.setLatLng({lng: lon_number, lat: lat_number});
+           } else
+              console.log('position zeroed');
+        } else
+           console.log('position undefined');
 
-        var lat_value = coordinates[0].substring(0, 11) + Array(11 - coordinates[0].substring(0, 11).length + 1).join(" ");
-        var lon_value = coordinates[1].substring(0, 11) + Array(11 - coordinates[1].substring(0, 11).length + 1).join(" ");
-
-        locMark.setLatLng({lng: Number(lon_value), lat: Number(lat_value)});
     });
 
     socket.on("current config rover", function(msg) {
